File: Project_and_Analysis_of_Algorithms/lab_graph_modeling/lab_graph_model.py
adj = {
    0: [1, 2],
    1: [0, 2],
    2: [0, 1],
    3: []
}

matriz = [
    [0, 0, 0, 0],
    [0, 0, 0, 0],
    [0, 0, 0, 0],
    [0, 0, 0, 0]
]

# Θ(n * deg(m))
def adj_list_to_adj_matrix(adj: dict[int, list[int]], matrix: list[list[int]]) -> list[list[int]]:
    """
        adj_list_to_adj_matrix returns the corresponding matrix from a given adjacency list representing a graph.
    """
    for k in adj.keys():
        for val in adj[k]:
            matrix[k][val] = 1
            matrix[val][k] = 1

    return matrix

for i in adj_list_to_adj_matrix(adj, matriz):
    print(i)

File: Project_and_Analysis_of_Algorithms/lab_graph_modeling/lab_graph_model.cpp
#include <bits/stdc++.h>
using namespace std;

void adj_list_to_adj_matrix(unordered_map<int, vector<int>> adj, vector<vector<int>> *matrix) {
    for(auto const&[node, neighbours] : adj) {
        for (int neighbour : neighbours) {
            (*matrix)[node][neighbour] = 1;
        }
    }
}

int main(void) {
    unordered_map<int, vector<int>> adj = {
        {0, {1, 2}},
        {1, {0, 2}},
        {2, {0, 1}},
        {3, {}}
    };

    vector<vector<int>> matrix = {
        {0, 0, 0, 0},
        {0, 0, 0, 0},
        {0, 0, 0, 0},
        {0, 0, 0, 0}
    };

    adj_list_to_adj_matrix(adj, &matrix);

    for(vector<int> i : matrix){
        for (int j : i) {
            cout << j << " ";
        }
        cout << "\n";
    }
}
